File: gallery/pptx/android/common/src/main/kotlin/fi/ranger/pptx/evg/EvgPainter.kt
package fi.ranger.pptx.evg

import fi.ranger.pptx.rgr.EVGDisplayList
import fi.ranger.pptx.rgr.EVGDrawCmd

/**
 * The walk: an `EVGDisplayList` in, calls on an [EvgSurface] out.
 *
 * The display list's own comment says a sixth backend should not be a sixth
 * copy of the traversal, and this is that rule applied one level down: Android
 * and AWT differ in how they fill a rounded box, not in what a command *means*,
 * so what a command means is decided exactly once, here.
 *
 * Two things this painter honours that the existing backends do not:
 *
 *  * **Gradients and shadows.** `EVGDisplayList.toJson` carries neither, so the
 *    WebGL host cannot draw them and only the CPU rasteriser does. Reading the
 *    objects directly — which is what a Kotlin host can do and a browser cannot
 *    — brings both back.
 *  * **Clipping and multi-ring paths.** The SoftCanvas painter ignores
 *    `PUSH_CLIP` / `POP_CLIP` and flattens every path to one polygon, so a
 *    ring-with-a-hole comes out solid. `save`/`clipRect`/`restore` and a real
 *    even-odd fill are one line each on both surfaces here.
 */
object EvgPainter {

    private const val RECT = 0
    private const val BORDER = 1
    private const val IMAGE = 2
    private const val TEXT = 3
    private const val PUSH_CLIP = 4
    private const val POP_CLIP = 5
    private const val PATH = 6
    private const val STROKE = 7

    /** Anything below this is invisible; drawing it is work for nothing. */
    private const val MIN_ALPHA = 0.001

    /**
     * Paint every command in order. Returns how many were actually drawn, which
     * is what a host puts in a debug overlay — a frame that draws 0 of 300
     * commands is a bug you want to see rather than a black screen.
     */
    fun paint(list: EVGDisplayList, surface: EvgSurface): Int {
        var drawn = 0
        // A list that pushes more clips than it pops must not leave the surface
        // with a clip nobody asked for, and one that pops more than it pushes
        // must not restore past the state the host handed us. The depth counter
        // is the whole of both guarantees.
        var clipDepth = 0
        val n = list.count()
        var i = 0
        while (i < n) {
            val c = list.at(i)
            i++
            when (c.kind) {
                PUSH_CLIP -> {
                    surface.save()
                    surface.clipRect(c.x.toFloat(), c.y.toFloat(), c.w.toFloat(), c.h.toFloat())
                    clipDepth++
                    continue
                }
                POP_CLIP -> {
                    if (clipDepth > 0) {
                        surface.restore()
                        clipDepth--
                    }
                    continue
                }
            }

            // A rotation turns the command about its own box centre, which is
            // what the PDF matrix and the GL shader both do. Wrapping it in
            // save/restore keeps the clip stack above untouched.
            val turned = c.rotate > 0.001 || c.rotate < -0.001
            if (turned) {
                surface.save()
                surface.rotate(
                    c.rotate.toFloat(),
                    (c.x + c.w * 0.5).toFloat(),
                    (c.y + c.h * 0.5).toFloat(),
                )
            }

            if (drawOne(c, surface)) drawn++

            if (turned) surface.restore()
        }
        while (clipDepth > 0) {
            surface.restore()
            clipDepth--
        }
        return drawn
    }

    private fun drawOne(c: EVGDrawCmd, surface: EvgSurface): Boolean {
        when (c.kind) {
            RECT -> {
                val shadow = shadowOf(c)
                val gradient = gradientOf(c)
                if (c.a < MIN_ALPHA && gradient == null && shadow == null) return false
                surface.fillRect(
                    c.x.toFloat(), c.y.toFloat(), c.w.toFloat(), c.h.toFloat(),
                    c.radius.toFloat(),
                    colorOf(c),
                    gradient,
                    shadow,
                )
                return true
            }
            BORDER -> {
                if (c.a < MIN_ALPHA) return false
                // A border with no thickness is still a border: the display
                // list leaves the default at 0 for a hairline, and a stroke of
                // width 0 is invisible on both surfaces.
                val t = if (c.thickness > 0.0) c.thickness.toFloat() else 1f
                surface.strokeRect(
                    c.x.toFloat(), c.y.toFloat(), c.w.toFloat(), c.h.toFloat(),
                    c.radius.toFloat(), t,
                    colorOf(c),
                )
                return true
            }
            IMAGE -> {
                if (c.src.isEmpty()) return false
                surface.drawImage(
                    c.src,
                    c.x.toFloat(), c.y.toFloat(), c.w.toFloat(), c.h.toFloat(),
                    c.radius.toFloat(),
                    c.flipH, c.flipV,
                )
                return true
            }
            TEXT -> {
                if (c.text.isEmpty()) return false
                if (c.a < MIN_ALPHA) return false
                surface.drawTextRun(
                    c.text,
                    c.x.toFloat(), c.y.toFloat(),
                    c.fontSize.toFloat(),
                    c.fontFamily,
                    isBold(c.fontWeight),
                    // `addText` has nowhere else to record a slanted face, so
                    // it writes "italic" into textAlign and `toJson` reads it
                    // back out the same way. Not lovely, but it is the format.
                    c.textAlign == "italic",
                    colorOf(c),
                )
                return true
            }
            PATH -> {
                if (c.a < MIN_ALPHA) return false
                if (c.pts.size < 6) return false
                surface.fillPath(c.pts, ringsOf(c), c.evenOdd, colorOf(c), shadowOf(c))
                return true
            }
            STROKE -> {
                if (c.a < MIN_ALPHA) return false
                if (c.pts.size < 4) return false
                val t = if (c.thickness > 0.0) c.thickness.toFloat() else 1f
                surface.strokePath(c.pts, ringsOf(c), t, colorOf(c))
                return true
            }
        }
        return false
    }

    /**
     * A path emitted by the SVG walk carries its ring boundaries; one emitted
     * by `addPolygon` / `addPolyline` may not. A single implicit ring covering
     * every point is the right reading of an empty `ringEnds` — it is what the
     * polygon filler assumes.
     */
    private fun ringsOf(c: EVGDrawCmd): List<Int> =
        if (c.ringEnds.isEmpty()) listOf(c.pts.size) else c.ringEnds

    private fun gradientOf(c: EVGDrawCmd): EvgGradient? {
        if (!c.hasGrad) return null
        return EvgGradient(
            color1 = argb(c.a, c.r, c.g, c.b),
            color2 = argb(c.a2, c.r2, c.g2, c.b2),
            // dir 0 = vertical (top to bottom), 1 = horizontal (left to right)
            vertical = c.gradDir != 1,
        )
    }

    private fun shadowOf(c: EVGDrawCmd): EvgShadow? {
        if (!c.hasShadow) return null
        if (c.shadowA < MIN_ALPHA) return null
        return EvgShadow(
            dx = c.shadowX.toFloat(),
            dy = c.shadowY.toFloat(),
            blur = c.shadowBlur.toFloat(),
            color = argb(c.shadowA, c.shadowR, c.shadowG, c.shadowB),
        )
    }

    private fun colorOf(c: EVGDrawCmd): Int = argb(c.a, c.r, c.g, c.b)

    /** 0-255 channels and a 0-1 alpha, the way every command carries them. */
    fun argb(alpha: Double, r: Int, g: Int, b: Int): Int {
        val a = clamp255((alpha * 255.0 + 0.5).toInt())
        return (a shl 24) or (clamp255(r) shl 16) or (clamp255(g) shl 8) or clamp255(b)
    }

    private fun clamp255(v: Int): Int = if (v < 0) 0 else if (v > 255) 255 else v

    /**
     * The display list writes the word, not the number — `addText` only ever
     * says "bold" — but a stylesheet path could put a CSS weight here, so a
     * number is read as one.
     */
    private fun isBold(weight: String): Boolean {
        if (weight.isEmpty()) return false
        if (weight == "bold" || weight == "bolder") return true
        val n = weight.toIntOrNull() ?: return false
        return n >= 600
    }
}

File: gallery/evg/apple/Sources/EvgPainter.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// The walk: an `EVGDisplayList` in, calls on an `EvgSurface` out.
//
// Android and Apple differ in how they fill a rounded box, not in what a
// command *means*, so what a command means is decided exactly once — here —
// and this file is a transliteration of `gallery/evg/android`'s `EvgPainter`
// rather than a second reading of the format. Where the two disagree, one of
// them is wrong.
//
// Two things this painter honours that a JSON-fed backend cannot:
//
//  * **Gradients and shadows.** `EVGDisplayList.toJson` carries neither, so a
//    browser host cannot draw them. Reading the objects directly — which is
//    what a Swift host can do — brings both back.
//  * **Clipping and multi-ring paths.** `save`/`clipRect`/`restore` and a real
//    even-odd fill are one line each on CoreGraphics.

import CoreGraphics

enum EvgPainter {

    private static let RECT = 0
    private static let BORDER = 1
    private static let IMAGE = 2
    private static let TEXT = 3
    private static let PUSH_CLIP = 4
    private static let POP_CLIP = 5
    private static let PATH = 6
    private static let STROKE = 7

    /// Anything below this is invisible; drawing it is work for nothing.
    private static let MIN_ALPHA = 0.001

    /// Paint every command in order. Answers how many were actually drawn,
    /// which is what a host puts in a debug overlay — a frame that draws 0 of
    /// 350 commands is a bug you want to see rather than a black screen.
    @discardableResult
    static func paint(_ list: EVGDisplayList, into surface: EvgSurface) -> Int {
        var drawn = 0
        // A list that pushes more clips than it pops must not leave the surface
        // with a clip nobody asked for, and one that pops more than it pushes
        // must not restore past the state the host handed us. The depth counter
        // is the whole of both guarantees.
        var clipDepth = 0
        let n = list.count()
        var i = 0
        while i < n {
            let c = list.at(i: i)
            i += 1

            if c.kind == PUSH_CLIP {
                surface.save()
                surface.clipRect(
                    x: CGFloat(c.x), y: CGFloat(c.y),
                    w: CGFloat(c.w), h: CGFloat(c.h)
                )
                clipDepth += 1
                continue
            }
            if c.kind == POP_CLIP {
                if clipDepth > 0 {
                    surface.restore()
                    clipDepth -= 1
                }
                continue
            }

            // A rotation turns the command about its own box centre, which is
            // what the PDF matrix and the GL shader both do. Wrapping it in
            // save/restore keeps the clip stack above untouched.
            let turned = c.rotate > 0.001 || c.rotate < -0.001
            if turned {
                surface.save()
                surface.rotate(
                    degrees: CGFloat(c.rotate),
                    px: CGFloat(c.x + c.w * 0.5),
                    py: CGFloat(c.y + c.h * 0.5)
                )
            }

            if drawOne(c, surface) { drawn += 1 }

            if turned { surface.restore() }
        }
        while clipDepth > 0 {
            surface.restore()
            clipDepth -= 1
        }
        return drawn
    }

    private static func drawOne(_ c: EVGDrawCmd, _ surface: EvgSurface) -> Bool {
        switch c.kind {
        case RECT:
            let shadow = shadowOf(c)
            let gradient = gradientOf(c)
            if c.a < MIN_ALPHA && gradient == nil && shadow == nil { return false }
            surface.fillRect(
                x: CGFloat(c.x), y: CGFloat(c.y), w: CGFloat(c.w), h: CGFloat(c.h),
                radius: CGFloat(c.radius),
                color: colorOf(c),
                gradient: gradient,
                shadow: shadow
            )
            return true

        case BORDER:
            if c.a < MIN_ALPHA { return false }
            // A border with no thickness is still a border: the display list
            // leaves the default at 0 for a hairline, and a stroke of width 0
            // is invisible.
            let t = c.thickness > 0.0 ? CGFloat(c.thickness) : 1.0
            surface.strokeRect(
                x: CGFloat(c.x), y: CGFloat(c.y), w: CGFloat(c.w), h: CGFloat(c.h),
                radius: CGFloat(c.radius), thickness: t,
                color: colorOf(c)
            )
            return true

        case IMAGE:
            if c.src.isEmpty { return false }
            surface.drawImage(
                src: c.src,
                x: CGFloat(c.x), y: CGFloat(c.y), w: CGFloat(c.w), h: CGFloat(c.h),
                radius: CGFloat(c.radius),
                flipH: c.flipH, flipV: c.flipV
            )
            return true

        case TEXT:
            if c.text.isEmpty { return false }
            if c.a < MIN_ALPHA { return false }
            surface.drawTextRun(
                text: c.text,
                x: CGFloat(c.x), top: CGFloat(c.y),
                sizePx: CGFloat(c.fontSize),
                family: c.fontFamily,
                bold: isBold(c.fontWeight),
                // `addText` has nowhere else to record a slanted face, so it
                // writes "italic" into textAlign and `toJson` reads it back out
                // the same way. Not lovely, but it is the format.
                italic: c.textAlign == "italic",
                color: colorOf(c)
            )
            return true

        case PATH:
            if c.a < MIN_ALPHA { return false }
            if c.pts.count < 6 { return false }
            surface.fillPath(
                pts: c.pts, ringEnds: ringsOf(c), evenOdd: c.evenOdd,
                color: colorOf(c), shadow: shadowOf(c)
            )
            return true

        case STROKE:
            if c.a < MIN_ALPHA { return false }
            if c.pts.count < 4 { return false }
            let t = c.thickness > 0.0 ? CGFloat(c.thickness) : 1.0
            surface.strokePath(
                pts: c.pts, ringEnds: ringsOf(c), thickness: t, color: colorOf(c)
            )
            return true

        default:
            return false
        }
    }

    /// A path emitted by the SVG walk carries its ring boundaries; one emitted
    /// by `addPolygon` / `addPolyline` may not. A single implicit ring covering
    /// every point is the right reading of an empty `ringEnds` — it is what the
    /// polygon filler assumes.
    private static func ringsOf(_ c: EVGDrawCmd) -> [Int] {
        c.ringEnds.isEmpty ? [c.pts.count] : c.ringEnds
    }

    private static func gradientOf(_ c: EVGDrawCmd) -> EvgGradient? {
        if !c.hasGrad { return nil }
        return EvgGradient(
            color1: EvgColor(r: c.r, g: c.g, b: c.b, a: c.a),
            color2: EvgColor(r: c.r2, g: c.g2, b: c.b2, a: c.a2),
            // dir 0 = vertical (top to bottom), 1 = horizontal (left to right)
            vertical: c.gradDir != 1
        )
    }

    private static func shadowOf(_ c: EVGDrawCmd) -> EvgShadow? {
        if !c.hasShadow { return nil }
        if c.shadowA < MIN_ALPHA { return nil }
        return EvgShadow(
            dx: CGFloat(c.shadowX),
            dy: CGFloat(c.shadowY),
            blur: CGFloat(c.shadowBlur),
            color: EvgColor(r: c.shadowR, g: c.shadowG, b: c.shadowB, a: c.shadowA)
        )
    }

    private static func colorOf(_ c: EVGDrawCmd) -> EvgColor {
        EvgColor(r: c.r, g: c.g, b: c.b, a: c.a)
    }

    /// The display list writes the word, not the number — `addText` only ever
    /// says "bold" — but a stylesheet path could put a CSS weight here, so a
    /// number is read as one.
    static func isBold(_ weight: String) -> Bool {
        if weight.isEmpty { return false }
        if weight == "bold" || weight == "bolder" { return true }
        guard let n = Int(weight) else { return false }
        return n >= 600
    }
}
